tempesta_fw/sock.c

```diff
@@ -419,30 +416,30 @@ ss_droplink(struct sock *sk)
 }
 
 /**
- * The function should be called with @sync = true whenever possible to
- * improve performance. If @sync = false, then return value must be checked
+ * The function should be called with SS_F_SYNC flag whenever possible to
+ * improve performance. Without SS_F_SYNC the return value must be checked
  * and the call must be repeated in case of bad return value.
- * @sync = true doesn't mean that the socket will be closed immediately,
+ * Note, that SS_F_SYNC doesn't mean that the socket will be closed immediately,
  * but rather it guarantees that the socket will be closed and the caller can
  * not care about return value.
  */
 int
-__ss_close(struct sock *sk, bool sync, bool need_drop)
+__ss_close(struct sock *sk, int flags)
 {
-
 	if (unlikely(!sk))
 		return SS_OK;
 	sk_incoming_cpu_update(sk);
 
-	if (!sync || !in_softirq()
+	if (!(flags & SS_F_SYNC) || !in_softirq()
 	    || smp_processor_id() != sk->sk_incoming_cpu)
 	{
 		SsWork sw = {
 			.sk	= sk,
+			.flags  = flags,
 			.action	= SS_CLOSE,
 		};
 
-		return ss_wq_push(&sw, sync);
+		return ss_wq_push(&sw, false);
```

Comment:
If we pass here e.g. by `!in_softirq()`, then `flags & SS_F_SYNC` can be true. I.e. we need to put the work to the queue, but do this synchronously.